pkg/pubsub/adapter/adapter.go

```diff
@@ -18,253 +18,202 @@ package adapter
-		if a.outbound, err = a.newHTTPClient(ctx, a.Sink); err != nil {
-			return fmt.Errorf("failed to create outbound cloudevent client: %w", err)
-		}
+	// Stop either if the adapter stops (sending to errCh) or if the context Done channel is closed.
+	select {
+	case err := <-errCh:
+		return err
+	case <-ctx.Done():
+		break
 	}
 
-	if a.reporter == nil {
-		a.reporter = NewStatsReporter()
+	// Done channel has been closed, we need to gracefully shutdown. The cancel() method will start its
+	// shutdown, if it hasn't finished in a reasonable amount of time, just return an error.
+	cancel()
+	select {
+	case err := <-errCh:
+		return err
+	case <-time.After(30 * time.Second):
+		return errors.New("timeout shutting down adapter")
 	}
+}
 
-	// Make the transformer client in case the TransformerURI has been set.
-	if a.Transformer != "" {
-		if a.transformer == nil {
-			if a.transformer, err = kncloudevents.NewDefaultClient(a.Transformer); err != nil {
-				return fmt.Errorf("failed to create transformer cloudevent client: %w", err)
-			}
-		}
+func (a *Adapter) receive(ctx context.Context, msg *pubsub.Message) {
+	event, err := converters.Convert(ctx, msg, a.adapterType)
+	if err != nil {
+		a.logger.Debug("Failed to convert received message to an event, check the msg format: %w", zap.Error(err))
+		// Ack the message so it won't be retried
+		msg.Ack()
+		return
 	}
 
-	return a.inbound.StartReceiver(ctx, a.receive)
-}
-
-func (a *Adapter) receive(ctx context.Context, event cloudevents.Event, resp *cloudevents.EventResponse) error {
-	logger := logging.FromContext(ctx).With(zap.Any("event.id", event.ID()), zap.Any("sink", a.Sink))
+	// Apply CloudEvent override extensions to the outbound event.
+	for k, v := range a.extensions {
+		event.SetExtension(k, v)
+	}
 
-	// TODO Name and ResourceGroup might cause problems in the near future, as we might use a single receive-adapter
-	//  for multiple source objects. Same with Namespace, when doing multi-tenancy.
 	args := &ReportArgs{
-		Name:          a.Name,
-		Namespace:     a.Namespace,
-		EventType:     event.Type(),
-		EventSource:   event.Source(),
-		ResourceGroup: a.ResourceGroup,
+		EventType:   event.Type(),
+		EventSource: event.Source(),
 	}
 
-	var err error
+	message := (*binding.EventMessage)(event)
+
 	// If a transformer has been configured, then transform the message.
 	// Note that this path in the code will be executed when using the receive adapter as part of the underlying Channel
 	// of a Broker. We currently set the TransformerURI to be the address of the Broker filter pod.
-	// TODO consider renaming transformer as it is confusing.
-	if a.transformer != nil {
-		transformedCTX, transformedEvent, err := a.transformer.Send(ctx, event)
-		rtctx := cloudevents.HTTPTransportContextFrom(transformedCTX)
+	// TODO consider renaming transformer as it is confusing. And see if we can get rid of this.
+	if a.transformerURI != "" {
+		resp, err := a.sendMsg(ctx, a.transformerURI, message)
```

Comment:
Are we really hitting this code path?

Reply:
yes, for for replies we are... it's pretty nasty. Added a couple of TODOs in the code.. This has been here forever and we should clean it up